From bf9a7d670427a9f1172178061c660b67d21a10a1 Mon Sep 17 00:00:00 2001
From: JOSEF OUANO <cvlaidmir@gmail.com>
Date: Fri, 5 Apr 2024 06:33:13 +0000
Subject: [PATCH 1/6] wwan: Added initial eps bearer settings (LTE) to APN
 dialog

Allows for setting APN, Username, Password, Allowed Auth, APN Type
Setting.

Co-authored-by: Felipe Borges <felipeborges@gnome.org>
(cherry picked from commit bb3030621d9803c12266ab7bb0f571cc34469559)
---
 panels/wwan/cc-wwan-apn-dialog.c  |  44 ++-
 panels/wwan/cc-wwan-apn-dialog.ui |  75 +++-
 panels/wwan/cc-wwan-data.c        | 575 +++++++++++++++++++++++++++++-
 panels/wwan/cc-wwan-data.h        |  25 ++
 4 files changed, 699 insertions(+), 20 deletions(-)

diff --git a/panels/wwan/cc-wwan-apn-dialog.c b/panels/wwan/cc-wwan-apn-dialog.c
index 46bce8fa2..0fa8bbdea 100644
--- a/panels/wwan/cc-wwan-apn-dialog.c
+++ b/panels/wwan/cc-wwan-apn-dialog.c
@@ -54,6 +54,8 @@ struct _CcWwanApnDialog
   GtkCheckButton    *apn_radio_button;
   GtkScrolledWindow *apn_list_view;
   GtkStack          *apn_settings_stack;
+  GtkDropDown       *lte_auth_combo;
+  GtkDropDown       *lte_attach_combo;
 
   CcWwanData        *wwan_data;
   CcWwanDataApn     *apn_to_save;   /* The APN currently being edited */
@@ -139,6 +141,7 @@ cc_wwan_apn_add_clicked_cb (CcWwanApnDialog *self)
 
   gtk_widget_set_visible (GTK_WIDGET (self->add_button), FALSE);
   gtk_widget_set_visible (GTK_WIDGET (self->save_button), TRUE);
+
   self->apn_to_save = NULL;
   gtk_stack_set_visible_child (self->apn_settings_stack,
                                GTK_WIDGET (self->apn_edit_view));
@@ -164,6 +167,19 @@ cc_wwan_apn_save_clicked_cb (CcWwanApnDialog *self)
   cc_wwan_data_apn_set_username (apn, gtk_editable_get_text (GTK_EDITABLE (self->username_entry)));
   cc_wwan_data_apn_set_password (apn, gtk_editable_get_text (GTK_EDITABLE (self->password_entry)));
 
+  if (gtk_drop_down_get_selected (self->lte_attach_combo) == 1)
+    {
+      cc_wwan_data_apn_set_name (apn, name);
+      cc_wwan_data_apn_set_apn (apn, apn_name);
+
+      cc_wwan_data_apn_set_initial_eps_apn (apn, apn_name);
+      cc_wwan_data_apn_set_initial_eps_username (apn, gtk_editable_get_text (GTK_EDITABLE (self->username_entry)));
+      cc_wwan_data_apn_set_initial_eps_password (apn, gtk_editable_get_text (GTK_EDITABLE (self->password_entry)));
+      cc_wwan_data_apn_set_initial_eps_auth (apn, gtk_drop_down_get_selected (self->lte_auth_combo));
+    }
+
+  cc_wwan_data_apn_set_initial_eps_apntype (apn, gtk_drop_down_get_selected (self->lte_attach_combo));
+
   cc_wwan_data_save_apn (self->wwan_data, apn, NULL, NULL, NULL);
 
   gtk_widget_set_visible (GTK_WIDGET (self->save_button), FALSE);
@@ -229,6 +245,7 @@ cc_wwan_apn_edit_clicked_cb (CcWwanApnDialog *self,
   CcWwanDataApn *apn;
   CcWwanApnRow *row;
   GtkWidget *widget;
+  const gchar *apn_name, *username, *password;
 
   widget = gtk_widget_get_ancestor (GTK_WIDGET (button), CC_TYPE_WWAN_APN_ROW);
   row = CC_WWAN_APN_ROW (widget);
@@ -239,9 +256,28 @@ cc_wwan_apn_edit_clicked_cb (CcWwanApnDialog *self,
   gtk_widget_set_visible (GTK_WIDGET (self->add_button), FALSE);
 
   gtk_editable_set_text (GTK_EDITABLE (self->name_entry), cc_wwan_data_apn_get_name (apn));
-  gtk_editable_set_text (GTK_EDITABLE (self->apn_entry), cc_wwan_data_apn_get_apn (apn));
-  gtk_editable_set_text (GTK_EDITABLE (self->username_entry), cc_wwan_data_apn_get_username (apn));
-  gtk_editable_set_text (GTK_EDITABLE (self->password_entry), cc_wwan_data_apn_get_password (apn));
+  if (cc_wwan_data_apn_get_initial_eps_apntype (apn))
+    {
+      apn_name = cc_wwan_data_apn_get_initial_eps_apn (apn);
+      username = cc_wwan_data_apn_get_initial_eps_username (apn);
+      password = cc_wwan_data_apn_get_initial_eps_password (apn);
+
+      gtk_drop_down_set_selected (self->lte_auth_combo, cc_wwan_data_apn_get_initial_eps_auth (apn));
+      gtk_drop_down_set_selected (self->lte_attach_combo, 1);
+    }
+  else
+    {
+      apn_name = cc_wwan_data_apn_get_apn (apn);
+      username = cc_wwan_data_apn_get_username (apn);
+      password = cc_wwan_data_apn_get_password (apn);
+
+      gtk_drop_down_set_selected (self->lte_auth_combo, 0);
+      gtk_drop_down_set_selected (self->lte_attach_combo, 0);
+    }
+
+  gtk_editable_set_text (GTK_EDITABLE (self->apn_entry), apn_name ? apn_name : "");
+  gtk_editable_set_text (GTK_EDITABLE (self->username_entry), username ? username : "");
+  gtk_editable_set_text (GTK_EDITABLE (self->password_entry), password ? password : "");
 
   gtk_stack_set_visible_child (self->apn_settings_stack,
                                GTK_WIDGET (self->apn_edit_view));
@@ -393,6 +429,8 @@ cc_wwan_apn_dialog_class_init (CcWwanApnDialogClass *klass)
   gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, apn_radio_button);
   gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, apn_settings_stack);
   gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, back_button);
+  gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, lte_auth_combo);
+  gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, lte_attach_combo);
   gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, name_entry);
   gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, password_entry);
   gtk_widget_class_bind_template_child (widget_class, CcWwanApnDialog, save_button);
diff --git a/panels/wwan/cc-wwan-apn-dialog.ui b/panels/wwan/cc-wwan-apn-dialog.ui
index 7853756a2..558a5ffa3 100644
--- a/panels/wwan/cc-wwan-apn-dialog.ui
+++ b/panels/wwan/cc-wwan-apn-dialog.ui
@@ -2,7 +2,7 @@
 <interface>
   <requires lib="gtk" version="4.0"/>
   <template class="CcWwanApnDialog" parent="GtkDialog">
-    <property name="default-height">480</property>
+    <property name="default-height">510</property>
     <property name="default-width">360</property>
     <property name="hide-on-close">True</property>
     <property name="title" translatable="yes">Access Points</property>
@@ -193,6 +193,79 @@
                       </object>
                     </child>
 
+
+                    <!--Initial EPS Auth Type -->
+                    <child>
+                      <object class="GtkLabel">
+                        <property name="halign">end</property>
+                        <property name="valign">center</property>
+                        <property name="label" translatable="yes">Auth Type</property>
+                        <property name="mnemonic_widget">lte_auth_combo</property>
+                        <style>
+                          <class name="dim-label"/>
+                        </style>
+                        <layout>
+                          <property name="column">0</property>
+                          <property name="row">4</property>
+                        </layout>
+                      </object>
+                    </child>
+                    
+                    <child>
+                      <object class="GtkDropDown" id="lte_auth_combo">
+                        <property name="model">
+                          <object class="GtkStringList">
+                            <items>
+                              <item translatable="yes">None</item>
+                              <item translatable="yes">PAP</item>
+                              <item translatable="yes">CHAP</item>
+                              <item translatable="yes">EAP</item>
+                              <item translatable="yes">MSCHAP</item>
+                              <item translatable="yes">MSCHAPV2</item>
+                            </items>
+                          </object>
+                        </property>
+                        <layout>
+                          <property name="column">1</property>
+                          <property name="row">4</property>
+                        </layout>
+                      </object>
+                    </child>
+
+                    <!-- Initial EPS APN type -->
+                    <child>
+                      <object class="GtkLabel">
+                        <property name="halign">end</property>
+                        <property name="valign">center</property>
+                        <property name="label" translatable="yes">APN Type</property>
+                        <property name="mnemonic_widget">lte_attach_combo</property>
+                        <style>
+                          <class name="dim-label"/>
+                        </style>
+                        <layout>
+                          <property name="column">0</property>
+                          <property name="row">5</property>
+                        </layout>
+                      </object>
+                    </child>
+                    
+                    <child>
+                      <object class="GtkDropDown" id="lte_attach_combo">
+                        <property name="model">
+                          <object class="GtkStringList">
+                            <items>
+                              <item translatable="yes">Internet</item>
+                              <item translatable="yes">Attach</item>
+                            </items>
+                          </object>
+                        </property>
+                        <layout>
+                          <property name="column">1</property>
+                          <property name="row">5</property>
+                        </layout>
+                      </object>
+                    </child>
+
                   </object>
                 </child>
               </object> <!-- ./AdwPreferencesGroup -->
diff --git a/panels/wwan/cc-wwan-data.c b/panels/wwan/cc-wwan-data.c
index 2608ab224..da7276f36 100644
--- a/panels/wwan/cc-wwan-data.c
+++ b/panels/wwan/cc-wwan-data.c
@@ -1207,7 +1207,6 @@ cc_wwan_data_apn_new (void)
 const gchar *
 cc_wwan_data_apn_get_name (CcWwanDataApn *apn)
 {
-  g_return_val_if_fail (CC_IS_WWAN_DATA_APN (apn), "");
 
   if (apn->remote_connection)
     return nm_connection_get_id (NM_CONNECTION (apn->remote_connection));
@@ -1215,7 +1214,7 @@ cc_wwan_data_apn_get_name (CcWwanDataApn *apn)
   if (apn->access_method)
     return nma_mobile_access_method_get_name (apn->access_method);
 
-  return "";
+  return NULL;
 }
 
 /**
@@ -1262,9 +1261,7 @@ cc_wwan_data_apn_set_name (CcWwanDataApn *apn,
 const gchar *
 cc_wwan_data_apn_get_apn (CcWwanDataApn *apn)
 {
-  const gchar *apn_name = "";
-
-  g_return_val_if_fail (CC_IS_WWAN_DATA_APN (apn), "");
+  const gchar *apn_name;
 
   if (apn->remote_connection)
     {
@@ -1278,7 +1275,7 @@ cc_wwan_data_apn_get_apn (CcWwanDataApn *apn)
       apn_name = nma_mobile_access_method_get_3gpp_apn (apn->access_method);
     }
 
-  return apn_name ? apn_name : "";
+  return apn_name;
 }
 
 /**
@@ -1305,7 +1302,7 @@ cc_wwan_data_apn_set_apn (CcWwanDataApn *apn,
   g_return_if_fail (apn_name != NULL);
   g_return_if_fail (*apn_name != '\0');
 
-  if (g_str_equal (cc_wwan_data_apn_get_apn (apn), apn_name))
+  if (g_strcmp0 (cc_wwan_data_apn_get_apn (apn), apn_name) == 0)
     return;
 
   apn->modified = TRUE;
@@ -1327,9 +1324,7 @@ cc_wwan_data_apn_set_apn (CcWwanDataApn *apn,
 const gchar *
 cc_wwan_data_apn_get_username (CcWwanDataApn *apn)
 {
-  const gchar *username = "";
-
-  g_return_val_if_fail (CC_IS_WWAN_DATA_APN (apn), "");
+  const gchar *username;
 
   if (apn->remote_connection)
     {
@@ -1343,7 +1338,7 @@ cc_wwan_data_apn_get_username (CcWwanDataApn *apn)
       username = nma_mobile_access_method_get_username (apn->access_method);
     }
 
-  return username ? username : "";
+  return username;
 }
 
 /**
@@ -1390,9 +1385,7 @@ cc_wwan_data_apn_set_username (CcWwanDataApn *apn,
 const gchar *
 cc_wwan_data_apn_get_password (CcWwanDataApn *apn)
 {
-  const gchar *password = "";
-
-  g_return_val_if_fail (CC_IS_WWAN_DATA_APN (apn), "");
+  const gchar *password;
 
   if (NM_IS_REMOTE_CONNECTION (apn->remote_connection))
     {
@@ -1409,7 +1402,7 @@ cc_wwan_data_apn_get_password (CcWwanDataApn *apn)
       if (error)
         {
           g_warning ("Error: %s", error->message);
-          return "";
+          return NULL;
         }
     }
 
@@ -1425,7 +1418,7 @@ cc_wwan_data_apn_get_password (CcWwanDataApn *apn)
       password = nma_mobile_access_method_get_password (apn->access_method);
     }
 
-  return password ? password : "";
+  return password;
 }
 
 /**
@@ -1461,6 +1454,556 @@ cc_wwan_data_apn_set_password (CcWwanDataApn *apn,
                 NULL);
 }
 
+/**
+ * cc_wwan_data_apn_get_initial_eps_apn:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the initial eps bearer APN of @apn
+ *
+ * Returns: (transfer none): The initial eps bearer APN of @apn
+ */
+const gchar *
+cc_wwan_data_apn_get_initial_eps_apn (CcWwanDataApn *apn)
+{
+  const gchar *apn_name;
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+
+    setting = nm_connection_get_setting_gsm (NM_CONNECTION (apn->remote_connection));
+    apn_name = nm_setting_gsm_get_initial_eps_apn (setting);
+  }
+  /*
+  else if (apn->access_method)
+    {
+      apn_name = nma_mobile_access_method_get_3gpp_apn (apn->access_method);
+    }
+  */
+
+  return apn_name;
+}
+
+/**
+ * cc_wwan_data_apn_set_initial_eps_apn:
+ * @apn: A #CcWwanDataApn
+ * @apn_name: The initial eps bearer apn to be used,
+ * should not be empty
+ *
+ * For LTE modems, set the initial eps bearer apn of @apn to @apn_name.
+ * @apn_name is usually a URL like 窶彳xample.com窶・or
+ * a simple string like 窶彿nternet窶・ *
+ * @apn is only modified, use @cc_wwan_data_save_apn()
+ * to save the changes.
+ */
+void
+cc_wwan_data_apn_set_initial_eps_apn (CcWwanDataApn *apn,
+                                      const gchar   *apn_name)
+{
+  NMConnection *connection;
+  NMSettingGsm *setting;
+
+  g_return_if_fail (CC_IS_WWAN_DATA_APN (apn));
+  g_return_if_fail (apn_name != NULL);
+  g_return_if_fail (*apn_name != '\0');
+
+  if (g_strcmp0 (cc_wwan_data_apn_get_initial_eps_apn (apn), apn_name) == 0)
+    return;
+
+  apn->modified = TRUE;
+  connection = wwan_data_get_nm_connection (apn);
+  setting = nm_connection_get_setting_gsm (connection);
+  g_object_set (G_OBJECT (setting),
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_APN, apn_name,
+                NULL);
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_username:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the initial eps bearer username of @apn
+ *
+ * Returns: (transfer none): The initial eps bearer username of @apn
+ */
+const gchar *
+cc_wwan_data_apn_get_initial_eps_username (CcWwanDataApn *apn)
+{
+  const gchar *username;
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+
+    setting = nm_connection_get_setting_gsm (NM_CONNECTION (apn->remote_connection));
+    username = nm_setting_gsm_get_initial_eps_username (setting);
+  }
+
+  return username;
+}
+
+/**
+ * cc_wwan_data_apn_set_initial_eps_username:
+ * @apn: A #CcWwanDataAPN
+ * @username: The initial eps bearer username to be used
+ *
+ * For LTE modems, set the initial eps bearer username of @apn to @username.
+ *
+ * @apn is only modified, use @cc_wwan_data_save_apn()
+ * to save the changes.
+ */
+void
+cc_wwan_data_apn_set_initial_eps_username (CcWwanDataApn *apn,
+                               const gchar   *username)
+{
+  NMConnection *connection;
+  NMSettingGsm *setting;
+
+  g_return_if_fail (CC_IS_WWAN_DATA_APN (apn));
+
+  if (username && !*username)
+    username = NULL;
+
+  if (g_strcmp0 (cc_wwan_data_apn_get_initial_eps_username (apn), username) == 0)
+    return;
+
+  apn->modified = TRUE;
+  connection = wwan_data_get_nm_connection (apn);
+  setting = nm_connection_get_setting_gsm (connection);
+  g_object_set (G_OBJECT (setting),
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_USERNAME, username,
+                NULL);
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_password:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the initial eps bearer password of @apn
+ *
+ * Returns: (transfer none): The initial eps bearer password of @apn
+ */
+const gchar *
+cc_wwan_data_apn_get_initial_eps_password (CcWwanDataApn *apn)
+{
+  const gchar *password;
+
+  if (NM_IS_REMOTE_CONNECTION (apn->remote_connection))
+    {
+      g_autoptr(GVariant) secrets = NULL;
+      g_autoptr(GError) error = NULL;
+
+      secrets = nm_remote_connection_get_secrets (NM_REMOTE_CONNECTION (apn->remote_connection),
+                                                  "gsm", NULL, &error);
+
+      if (!error)
+        nm_connection_update_secrets (NM_CONNECTION (apn->remote_connection),
+                                      "gsm", secrets, &error);
+
+      if (error)
+        {
+          g_warning ("Error: %s", error->message);
+          return NULL;
+        }
+    }
+
+  if (apn->remote_connection)
+    {
+      NMSettingGsm *setting;
+
+      setting = nm_connection_get_setting_gsm (NM_CONNECTION (apn->remote_connection));
+      password = nm_setting_gsm_get_initial_eps_password (setting);
+    }
+  /*
+  else if (apn->access_method)
+    {
+      password = nma_mobile_access_method_get_password (apn->access_method);
+    }
+  */
+
+  return password;
+}
+
+/**
+ * cc_wwan_data_apn_set_initial_eps_password:
+ * @apn: A #CcWwanDataApn
+ * @password: The initial eps bearer password to be used
+ *
+ * For LTE modems, get the initial eps bearer password of @apn to @password.
+ *
+ * @apn is only modified, use @cc_wwan_data_save_apn()
+ * to save the changes.
+ */
+void
+cc_wwan_data_apn_set_initial_eps_password (CcWwanDataApn *apn,
+                               const gchar   *password)
+{
+  NMConnection *connection;
+  NMSettingGsm *setting;
+
+  g_return_if_fail (CC_IS_WWAN_DATA_APN (apn));
+
+  if (password && !*password)
+    password = NULL;
+
+  if (g_strcmp0 (cc_wwan_data_apn_get_initial_eps_password (apn), password) == 0)
+    return;
+
+  apn->modified = TRUE;
+  connection = wwan_data_get_nm_connection (apn);
+  setting = nm_connection_get_setting_gsm (connection);
+  g_object_set (G_OBJECT (setting),
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_PASSWORD, password,
+                NULL);
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_noauth:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the noauth authentication status of @apn
+ *
+ * Returns: (transfer none): The noauth authentication status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_noauth (CcWwanDataApn *apn)
+{
+  gboolean noauth_enabled;
+
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
+    noauth_enabled = nm_setting_gsm_get_initial_eps_noauth(setting);
+  }
+  else if (apn->access_method)
+  {
+    return FALSE;
+  }
+  return noauth_enabled;
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_eap:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the eap authentication status of @apn
+ *
+ * Returns: (transfer none): The eap authentication status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_refuse_eap (CcWwanDataApn *apn)
+{
+  gboolean eap_enabled;
+
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
+    eap_enabled = nm_setting_gsm_get_initial_eps_refuse_eap(setting);
+  }
+  else if (apn->access_method)
+  {
+    return FALSE;
+  }
+  return eap_enabled;
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_pap:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the pap authentication status of @apn
+ *
+ * Returns: (transfer none): The pap authentication status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_refuse_pap (CcWwanDataApn *apn)
+{
+  gboolean pap_enabled;
+
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
+    pap_enabled = nm_setting_gsm_get_initial_eps_refuse_pap(setting);
+  }
+  else if (apn->access_method)
+  {
+    return FALSE;
+  }
+  return pap_enabled;
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_refuse_chap:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the chap authentication status of @apn
+ *
+ * Returns: (transfer none): The chap authentication status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_refuse_chap (CcWwanDataApn *apn)
+{
+  gboolean chap_enabled;
+
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
+    chap_enabled = nm_setting_gsm_get_initial_eps_refuse_chap(setting);
+  }
+  else if (apn->access_method)
+  {
+    return TRUE;
+  }
+  return chap_enabled;
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_refuse_mschap:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the mschap authentication status of @apn
+ *
+ * Returns: (transfer none): The mschap authentication status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_refuse_mschap (CcWwanDataApn *apn)
+{
+  gboolean mschap_enabled;
+
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
+    mschap_enabled = nm_setting_gsm_get_initial_eps_refuse_mschap(setting);
+  }
+  else if (apn->access_method)
+  {
+    return FALSE;
+  }
+  return mschap_enabled;
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_refuse_mschapv2:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the mschapv2 authentication status of @apn
+ *
+ * Returns: (transfer none): The mschapv2 authentication status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_refuse_mschapv2 (CcWwanDataApn *apn)
+{
+  gboolean mschapv2_enabled;
+
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
+    mschapv2_enabled = nm_setting_gsm_get_initial_eps_refuse_mschapv2(setting);
+  }
+  else if (apn->access_method)
+  {
+    return FALSE;
+  }
+  return mschapv2_enabled;
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_auth:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the authentication status for all auth types of @apn
+ *
+ * Returns: (transfer none): The authentication status of @apn
+ */
+guint
+cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn)
+{
+  if(cc_wwan_data_apn_get_initial_eps_noauth(apn) == TRUE) return 0;
+  else if(cc_wwan_data_apn_get_initial_eps_refuse_pap(apn) == FALSE) return 1;
+  else if(cc_wwan_data_apn_get_initial_eps_refuse_chap(apn) == FALSE) return 2;
+  else if(cc_wwan_data_apn_get_initial_eps_refuse_eap(apn) == FALSE) return 3;
+  else if(cc_wwan_data_apn_get_initial_eps_refuse_mschap(apn) == FALSE) return 4;
+  else if(cc_wwan_data_apn_get_initial_eps_refuse_mschapv2(apn) == FALSE) return 5;
+  else return 0;
+}
+
+/**
+ * cc_wwan_data_apn_set_initial_eps_auth:
+ * @apn: A #CcWwanDataApn
+ * @authtype: The initial eps bearer auth type
+ *
+ * For LTE modems, set the authentication type of @apn to @authtype.
+ *
+ * @apn is only modified, use @cc_wwan_data_save_apn()
+ * to save the changes.
+ */
+void
+cc_wwan_data_apn_set_initial_eps_auth ( CcWwanDataApn *apn,
+                            guint authtype)
+{
+  NMSettingGsm *setting;
+  NMConnection *connection;
+
+  gboolean noauth_enabled;
+  gboolean refuse_pap_enabled;
+  gboolean refuse_chap_enabled;
+  gboolean refuse_eap_enabled;
+  gboolean refuse_mschap_enabled;
+  gboolean refuse_mschapv2_enabled;
+
+  g_return_if_fail (CC_IS_WWAN_DATA_APN (apn));
+  apn->modified = TRUE;
+
+  connection = wwan_data_get_nm_connection (apn);
+  setting = nm_connection_get_setting_gsm (connection);
+
+  switch(authtype)
+  {
+    case 0: //None
+      noauth_enabled = TRUE;
+      refuse_pap_enabled = TRUE;
+      refuse_chap_enabled = TRUE;
+      refuse_eap_enabled = TRUE;
+      refuse_mschap_enabled = TRUE;
+      refuse_mschapv2_enabled = TRUE;
+      break;
+    case 1: //PAP
+      noauth_enabled = FALSE;
+      refuse_pap_enabled = FALSE;
+      refuse_chap_enabled = TRUE;
+      refuse_eap_enabled = TRUE;
+      refuse_mschap_enabled = TRUE;
+      refuse_mschapv2_enabled = TRUE;
+      break;
+    case 2: //CHAP
+      noauth_enabled = FALSE;
+      refuse_pap_enabled = TRUE;
+      refuse_chap_enabled = FALSE;
+      refuse_eap_enabled = TRUE;
+      refuse_mschap_enabled = TRUE;
+      refuse_mschapv2_enabled = TRUE;
+      break;
+    case 3: //EAP
+      noauth_enabled = FALSE;
+      refuse_pap_enabled = TRUE;
+      refuse_chap_enabled = TRUE;
+      refuse_eap_enabled = FALSE;
+      refuse_mschap_enabled = TRUE;
+      refuse_mschapv2_enabled = TRUE;
+      break;
+    case 4: //MSCHAP
+      noauth_enabled = FALSE;
+      refuse_pap_enabled = TRUE;
+      refuse_chap_enabled = TRUE;
+      refuse_eap_enabled = TRUE;
+      refuse_mschap_enabled = FALSE;
+      refuse_mschapv2_enabled = TRUE;
+      break;
+    case 5: //MSCHAPV2
+      noauth_enabled = FALSE;
+      refuse_pap_enabled = TRUE;
+      refuse_chap_enabled = TRUE;
+      refuse_eap_enabled = TRUE;
+      refuse_mschap_enabled = TRUE;
+      refuse_mschapv2_enabled = FALSE;
+      break;
+  }
+
+	g_object_set (setting,
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_NOAUTH, noauth_enabled,
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_REFUSE_PAP, refuse_pap_enabled,
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_REFUSE_CHAP, refuse_chap_enabled,
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_REFUSE_EAP, refuse_eap_enabled,
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_REFUSE_MSCHAP, refuse_mschap_enabled,
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_REFUSE_MSCHAPV2, refuse_mschapv2_enabled,
+                NULL);
+}
+
+/**
+ * cc_wwan_data_apn_get_initial_eps_apntype:
+ * @apn: A #CcWwanDataApn
+ *
+ * For LTE modems, get the initial eps bearer configure of @apn
+ *
+ * Returns: (transfer none): The initial eps bearer configure status of @apn
+ */
+gboolean
+cc_wwan_data_apn_get_initial_eps_apntype (CcWwanDataApn *apn)
+{
+  gboolean apn_type = FALSE;
+
+  g_return_val_if_fail (CC_IS_WWAN_DATA_APN (apn), FALSE);
+
+  if (apn->remote_connection)
+  {
+    NMSettingGsm *setting;
+
+    setting = nm_connection_get_setting_gsm (NM_CONNECTION (apn->remote_connection));
+    apn_type = nm_setting_gsm_get_initial_eps_config (setting);
+  }
+  /*
+  else if (apn->access_method)
+    {
+      apn_name = nma_mobile_access_method_get_3gpp_apn (apn->access_method);
+    }
+  */
+
+  return apn_type;
+}
+
+/**
+ * cc_wwan_data_apn_set_initial_eps_apntype:
+ * @apn: A #CcWwanDataApn
+ * @apntype: The initial eps bearer configure to be used
+ *
+ * For LTE modems, set the apn type of @apn to @apntype.
+ *
+ * @apn is only modified, use @cc_wwan_data_save_apn()
+ * to save the changes.
+ */
+void
+cc_wwan_data_apn_set_initial_eps_apntype (CcWwanDataApn *apn,
+                              guint apntype)
+{
+  NMConnection *connection;
+  NMSettingGsm *setting;
+  gboolean init_eps_config = TRUE;
+
+
+  g_return_if_fail (CC_IS_WWAN_DATA_APN (apn));
+
+  apn->modified = TRUE;
+  connection = wwan_data_get_nm_connection(apn);
+  setting = nm_connection_get_setting_gsm(connection);
+
+  if(apntype == 0){
+    init_eps_config = FALSE;
+  }
+  else if(apntype == 1){
+    init_eps_config = TRUE;
+  }
+  else
+  {
+    init_eps_config = FALSE;
+  }
+
+  g_object_set (setting,
+              NM_SETTING_GSM_INITIAL_EPS_BEARER_CONFIGURE, init_eps_config,
+              NULL);
+}
+
 gint
 cc_wwan_data_get_priority (CcWwanData *self)
 {
diff --git a/panels/wwan/cc-wwan-data.h b/panels/wwan/cc-wwan-data.h
index 9572b862d..f39e264ed 100644
--- a/panels/wwan/cc-wwan-data.h
+++ b/panels/wwan/cc-wwan-data.h
@@ -89,5 +89,30 @@ void           cc_wwan_data_apn_set_password (CcWwanDataApn *apn,
 gint           cc_wwan_data_get_priority     (CcWwanData    *self);
 void           cc_wwan_data_set_priority     (CcWwanData    *self,
                                               int            priority);
+const gchar   *cc_wwan_data_apn_get_initial_eps_apn      (CcWwanDataApn *apn);
+void           cc_wwan_data_apn_set_initial_eps_apn      (CcWwanDataApn *apn,
+                                                          const gchar   *apn_name);
+const gchar   *cc_wwan_data_apn_get_initial_eps_username (CcWwanDataApn *apn);
+void           cc_wwan_data_apn_set_initial_eps_username (CcWwanDataApn *apn,
+                                                          const gchar   *username);
+const gchar   *cc_wwan_data_apn_get_initial_eps_password (CcWwanDataApn *apn);
+void           cc_wwan_data_apn_set_initial_eps_password (CcWwanDataApn *apn,
+                                                          const gchar   *password);
+
+gboolean cc_wwan_data_apn_get_initial_eps_noauth (CcWwanDataApn *apn);
+gboolean cc_wwan_data_apn_get_initial_eps_refuse_pap  (CcWwanDataApn *apn);
+gboolean cc_wwan_data_apn_get_initial_eps_refuse_chap (CcWwanDataApn *apn);
+gboolean cc_wwan_data_apn_get_initial_eps_refuse_eap  (CcWwanDataApn *apn);
+gboolean cc_wwan_data_apn_get_initial_eps_refuse_mschap (CcWwanDataApn *apn);
+gboolean cc_wwan_data_apn_get_initial_eps_refuse_mschapv2 (CcWwanDataApn *apn);
+
+guint cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn);
+void cc_wwan_data_apn_set_initial_eps_auth (CcWwanDataApn *apn,
+                                            guint allowed_auth);
+
+gboolean cc_wwan_data_apn_get_initial_eps_apntype (CcWwanDataApn *apn);
+
+void cc_wwan_data_apn_set_initial_eps_apntype (CcWwanDataApn *apn,
+                                               guint         apntype);
 
 G_END_DECLS
-- 
2.49.0


From b908b5064ffe8f08a79555dfd8fc89857d0e17a1 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 13 May 2025 13:22:25 +0200
Subject: [PATCH 2/6] wwan: Rename method that exposes
 initial-eps-bearer-configure

For the sake of clarity of what it does.

(cherry picked from commit 6d302afbecfeb92b8bbcbc7af85609b3c755c74f)
---
 panels/wwan/cc-wwan-apn-dialog.c |  4 ++--
 panels/wwan/cc-wwan-data.c       | 31 ++++++++++---------------------
 panels/wwan/cc-wwan-data.h       |  6 +++---
 3 files changed, 15 insertions(+), 26 deletions(-)

diff --git a/panels/wwan/cc-wwan-apn-dialog.c b/panels/wwan/cc-wwan-apn-dialog.c
index 0fa8bbdea..1922ef8e0 100644
--- a/panels/wwan/cc-wwan-apn-dialog.c
+++ b/panels/wwan/cc-wwan-apn-dialog.c
@@ -178,7 +178,7 @@ cc_wwan_apn_save_clicked_cb (CcWwanApnDialog *self)
       cc_wwan_data_apn_set_initial_eps_auth (apn, gtk_drop_down_get_selected (self->lte_auth_combo));
     }
 
-  cc_wwan_data_apn_set_initial_eps_apntype (apn, gtk_drop_down_get_selected (self->lte_attach_combo));
+  cc_wwan_data_apn_set_should_configure_initial_eps_bearer (apn, gtk_drop_down_get_selected (self->lte_attach_combo));
 
   cc_wwan_data_save_apn (self->wwan_data, apn, NULL, NULL, NULL);
 
@@ -256,7 +256,7 @@ cc_wwan_apn_edit_clicked_cb (CcWwanApnDialog *self,
   gtk_widget_set_visible (GTK_WIDGET (self->add_button), FALSE);
 
   gtk_editable_set_text (GTK_EDITABLE (self->name_entry), cc_wwan_data_apn_get_name (apn));
-  if (cc_wwan_data_apn_get_initial_eps_apntype (apn))
+  if (cc_wwan_data_apn_should_configure_initial_eps_bearer (apn))
     {
       apn_name = cc_wwan_data_apn_get_initial_eps_apn (apn);
       username = cc_wwan_data_apn_get_initial_eps_username (apn);
diff --git a/panels/wwan/cc-wwan-data.c b/panels/wwan/cc-wwan-data.c
index da7276f36..64837e570 100644
--- a/panels/wwan/cc-wwan-data.c
+++ b/panels/wwan/cc-wwan-data.c
@@ -1932,15 +1932,16 @@ cc_wwan_data_apn_set_initial_eps_auth ( CcWwanDataApn *apn,
 }
 
 /**
- * cc_wwan_data_apn_get_initial_eps_apntype:
+ * cc_wwan_data_apn_should_configure_initial_eps_bearer:
  * @apn: A #CcWwanDataApn
  *
- * For LTE modems, get the initial eps bearer configure of @apn
+ * For LTE modems, Determine whether NetworkManager should explicitly
+ * configure the initial EPS bearer.
  *
  * Returns: (transfer none): The initial eps bearer configure status of @apn
  */
 gboolean
-cc_wwan_data_apn_get_initial_eps_apntype (CcWwanDataApn *apn)
+cc_wwan_data_apn_should_configure_initial_eps_bearer (CcWwanDataApn *apn)
 {
   gboolean apn_type = FALSE;
 
@@ -1968,19 +1969,18 @@ cc_wwan_data_apn_get_initial_eps_apntype (CcWwanDataApn *apn)
  * @apn: A #CcWwanDataApn
  * @apntype: The initial eps bearer configure to be used
  *
- * For LTE modems, set the apn type of @apn to @apntype.
+ * For LTE modems, set whether NetworkManager should explicitly configure
+ * the initial EPS bearer.
  *
  * @apn is only modified, use @cc_wwan_data_save_apn()
  * to save the changes.
  */
 void
-cc_wwan_data_apn_set_initial_eps_apntype (CcWwanDataApn *apn,
-                              guint apntype)
+cc_wwan_data_apn_set_should_configure_initial_eps_bearer (CcWwanDataApn *apn,
+                                                          gboolean       configure)
 {
   NMConnection *connection;
   NMSettingGsm *setting;
-  gboolean init_eps_config = TRUE;
-
 
   g_return_if_fail (CC_IS_WWAN_DATA_APN (apn));
 
@@ -1988,20 +1988,9 @@ cc_wwan_data_apn_set_initial_eps_apntype (CcWwanDataApn *apn,
   connection = wwan_data_get_nm_connection(apn);
   setting = nm_connection_get_setting_gsm(connection);
 
-  if(apntype == 0){
-    init_eps_config = FALSE;
-  }
-  else if(apntype == 1){
-    init_eps_config = TRUE;
-  }
-  else
-  {
-    init_eps_config = FALSE;
-  }
-
   g_object_set (setting,
-              NM_SETTING_GSM_INITIAL_EPS_BEARER_CONFIGURE, init_eps_config,
-              NULL);
+                NM_SETTING_GSM_INITIAL_EPS_BEARER_CONFIGURE, configure,
+                NULL);
 }
 
 gint
diff --git a/panels/wwan/cc-wwan-data.h b/panels/wwan/cc-wwan-data.h
index f39e264ed..6d75a4fd0 100644
--- a/panels/wwan/cc-wwan-data.h
+++ b/panels/wwan/cc-wwan-data.h
@@ -110,9 +110,9 @@ guint cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn);
 void cc_wwan_data_apn_set_initial_eps_auth (CcWwanDataApn *apn,
                                             guint allowed_auth);
 
-gboolean cc_wwan_data_apn_get_initial_eps_apntype (CcWwanDataApn *apn);
+gboolean cc_wwan_data_apn_should_configure_initial_eps_bearer (CcWwanDataApn *apn);
 
-void cc_wwan_data_apn_set_initial_eps_apntype (CcWwanDataApn *apn,
-                                               guint         apntype);
+void cc_wwan_data_apn_set_should_configure_initial_eps_bearer (CcWwanDataApn *apn,
+                                                               gboolean       configure);
 
 G_END_DECLS
-- 
2.49.0


From beba2278cb720c85e1777af38c31b643aba9e934 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 13 May 2025 13:34:23 +0200
Subject: [PATCH 3/6] wwan: Drop unused CcWwanData public headers

These are only used internally.

(cherry picked from commit 69965182efcc08575d877807ebcbbbb9b1045f37)
---
 panels/wwan/cc-wwan-data.c | 12 ++++++------
 panels/wwan/cc-wwan-data.h |  7 -------
 2 files changed, 6 insertions(+), 13 deletions(-)

diff --git a/panels/wwan/cc-wwan-data.c b/panels/wwan/cc-wwan-data.c
index 64837e570..5c37cc624 100644
--- a/panels/wwan/cc-wwan-data.c
+++ b/panels/wwan/cc-wwan-data.c
@@ -1665,7 +1665,7 @@ cc_wwan_data_apn_set_initial_eps_password (CcWwanDataApn *apn,
  *
  * Returns: (transfer none): The noauth authentication status of @apn
  */
-gboolean
+static gboolean
 cc_wwan_data_apn_get_initial_eps_noauth (CcWwanDataApn *apn)
 {
   gboolean noauth_enabled;
@@ -1692,7 +1692,7 @@ cc_wwan_data_apn_get_initial_eps_noauth (CcWwanDataApn *apn)
  *
  * Returns: (transfer none): The eap authentication status of @apn
  */
-gboolean
+static gboolean
 cc_wwan_data_apn_get_initial_eps_refuse_eap (CcWwanDataApn *apn)
 {
   gboolean eap_enabled;
@@ -1719,7 +1719,7 @@ cc_wwan_data_apn_get_initial_eps_refuse_eap (CcWwanDataApn *apn)
  *
  * Returns: (transfer none): The pap authentication status of @apn
  */
-gboolean
+static gboolean
 cc_wwan_data_apn_get_initial_eps_refuse_pap (CcWwanDataApn *apn)
 {
   gboolean pap_enabled;
@@ -1746,7 +1746,7 @@ cc_wwan_data_apn_get_initial_eps_refuse_pap (CcWwanDataApn *apn)
  *
  * Returns: (transfer none): The chap authentication status of @apn
  */
-gboolean
+static gboolean
 cc_wwan_data_apn_get_initial_eps_refuse_chap (CcWwanDataApn *apn)
 {
   gboolean chap_enabled;
@@ -1773,7 +1773,7 @@ cc_wwan_data_apn_get_initial_eps_refuse_chap (CcWwanDataApn *apn)
  *
  * Returns: (transfer none): The mschap authentication status of @apn
  */
-gboolean
+static gboolean
 cc_wwan_data_apn_get_initial_eps_refuse_mschap (CcWwanDataApn *apn)
 {
   gboolean mschap_enabled;
@@ -1800,7 +1800,7 @@ cc_wwan_data_apn_get_initial_eps_refuse_mschap (CcWwanDataApn *apn)
  *
  * Returns: (transfer none): The mschapv2 authentication status of @apn
  */
-gboolean
+static gboolean
 cc_wwan_data_apn_get_initial_eps_refuse_mschapv2 (CcWwanDataApn *apn)
 {
   gboolean mschapv2_enabled;
diff --git a/panels/wwan/cc-wwan-data.h b/panels/wwan/cc-wwan-data.h
index 6d75a4fd0..1539fc9d6 100644
--- a/panels/wwan/cc-wwan-data.h
+++ b/panels/wwan/cc-wwan-data.h
@@ -99,13 +99,6 @@ const gchar   *cc_wwan_data_apn_get_initial_eps_password (CcWwanDataApn *apn);
 void           cc_wwan_data_apn_set_initial_eps_password (CcWwanDataApn *apn,
                                                           const gchar   *password);
 
-gboolean cc_wwan_data_apn_get_initial_eps_noauth (CcWwanDataApn *apn);
-gboolean cc_wwan_data_apn_get_initial_eps_refuse_pap  (CcWwanDataApn *apn);
-gboolean cc_wwan_data_apn_get_initial_eps_refuse_chap (CcWwanDataApn *apn);
-gboolean cc_wwan_data_apn_get_initial_eps_refuse_eap  (CcWwanDataApn *apn);
-gboolean cc_wwan_data_apn_get_initial_eps_refuse_mschap (CcWwanDataApn *apn);
-gboolean cc_wwan_data_apn_get_initial_eps_refuse_mschapv2 (CcWwanDataApn *apn);
-
 guint cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn);
 void cc_wwan_data_apn_set_initial_eps_auth (CcWwanDataApn *apn,
                                             guint allowed_auth);
-- 
2.49.0


From f47de842c837d253e14a41ebefafc9847f4cb366 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 13 May 2025 14:03:36 +0200
Subject: [PATCH 4/6] wwan: Refactor resolution of EPS auth method

This still could use a enum, but lets leave this for later for when we
port the APN Auth Type from a GtkDropDown to AdwComboRow.

(cherry picked from commit 27c72bd2a2ad3119fde3b06039838d6fbe11ff0d)
---
 panels/wwan/cc-wwan-apn-dialog.c |   4 +-
 panels/wwan/cc-wwan-data.c       | 199 ++++---------------------------
 panels/wwan/cc-wwan-data.h       |   6 +-
 3 files changed, 31 insertions(+), 178 deletions(-)

diff --git a/panels/wwan/cc-wwan-apn-dialog.c b/panels/wwan/cc-wwan-apn-dialog.c
index 1922ef8e0..dfa6ea42f 100644
--- a/panels/wwan/cc-wwan-apn-dialog.c
+++ b/panels/wwan/cc-wwan-apn-dialog.c
@@ -175,7 +175,7 @@ cc_wwan_apn_save_clicked_cb (CcWwanApnDialog *self)
       cc_wwan_data_apn_set_initial_eps_apn (apn, apn_name);
       cc_wwan_data_apn_set_initial_eps_username (apn, gtk_editable_get_text (GTK_EDITABLE (self->username_entry)));
       cc_wwan_data_apn_set_initial_eps_password (apn, gtk_editable_get_text (GTK_EDITABLE (self->password_entry)));
-      cc_wwan_data_apn_set_initial_eps_auth (apn, gtk_drop_down_get_selected (self->lte_auth_combo));
+      cc_wwan_data_apn_set_initial_eps_auth_method (apn, gtk_drop_down_get_selected (self->lte_auth_combo));
     }
 
   cc_wwan_data_apn_set_should_configure_initial_eps_bearer (apn, gtk_drop_down_get_selected (self->lte_attach_combo));
@@ -262,7 +262,7 @@ cc_wwan_apn_edit_clicked_cb (CcWwanApnDialog *self,
       username = cc_wwan_data_apn_get_initial_eps_username (apn);
       password = cc_wwan_data_apn_get_initial_eps_password (apn);
 
-      gtk_drop_down_set_selected (self->lte_auth_combo, cc_wwan_data_apn_get_initial_eps_auth (apn));
+      gtk_drop_down_set_selected (self->lte_auth_combo, cc_wwan_data_apn_get_initial_eps_auth_method (apn));
       gtk_drop_down_set_selected (self->lte_attach_combo, 1);
     }
   else
diff --git a/panels/wwan/cc-wwan-data.c b/panels/wwan/cc-wwan-data.c
index 5c37cc624..b5efdebce 100644
--- a/panels/wwan/cc-wwan-data.c
+++ b/panels/wwan/cc-wwan-data.c
@@ -1658,189 +1658,42 @@ cc_wwan_data_apn_set_initial_eps_password (CcWwanDataApn *apn,
 }
 
 /**
- * cc_wwan_data_apn_get_initial_eps_noauth:
+ * cc_wwan_data_apn_get_initial_eps_auth_method:
  * @apn: A #CcWwanDataApn
  *
- * For LTE modems, get the noauth authentication status of @apn
- *
- * Returns: (transfer none): The noauth authentication status of @apn
- */
-static gboolean
-cc_wwan_data_apn_get_initial_eps_noauth (CcWwanDataApn *apn)
-{
-  gboolean noauth_enabled;
-
-
-  if (apn->remote_connection)
-  {
-    NMSettingGsm *setting;
-    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
-    noauth_enabled = nm_setting_gsm_get_initial_eps_noauth(setting);
-  }
-  else if (apn->access_method)
-  {
-    return FALSE;
-  }
-  return noauth_enabled;
-}
-
-/**
- * cc_wwan_data_apn_get_initial_eps_eap:
- * @apn: A #CcWwanDataApn
- *
- * For LTE modems, get the eap authentication status of @apn
- *
- * Returns: (transfer none): The eap authentication status of @apn
- */
-static gboolean
-cc_wwan_data_apn_get_initial_eps_refuse_eap (CcWwanDataApn *apn)
-{
-  gboolean eap_enabled;
-
-
-  if (apn->remote_connection)
-  {
-    NMSettingGsm *setting;
-    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
-    eap_enabled = nm_setting_gsm_get_initial_eps_refuse_eap(setting);
-  }
-  else if (apn->access_method)
-  {
-    return FALSE;
-  }
-  return eap_enabled;
-}
-
-/**
- * cc_wwan_data_apn_get_initial_eps_pap:
- * @apn: A #CcWwanDataApn
- *
- * For LTE modems, get the pap authentication status of @apn
- *
- * Returns: (transfer none): The pap authentication status of @apn
- */
-static gboolean
-cc_wwan_data_apn_get_initial_eps_refuse_pap (CcWwanDataApn *apn)
-{
-  gboolean pap_enabled;
-
-
-  if (apn->remote_connection)
-  {
-    NMSettingGsm *setting;
-    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
-    pap_enabled = nm_setting_gsm_get_initial_eps_refuse_pap(setting);
-  }
-  else if (apn->access_method)
-  {
-    return FALSE;
-  }
-  return pap_enabled;
-}
-
-/**
- * cc_wwan_data_apn_get_initial_eps_refuse_chap:
- * @apn: A #CcWwanDataApn
- *
- * For LTE modems, get the chap authentication status of @apn
- *
- * Returns: (transfer none): The chap authentication status of @apn
- */
-static gboolean
-cc_wwan_data_apn_get_initial_eps_refuse_chap (CcWwanDataApn *apn)
-{
-  gboolean chap_enabled;
-
-
-  if (apn->remote_connection)
-  {
-    NMSettingGsm *setting;
-    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
-    chap_enabled = nm_setting_gsm_get_initial_eps_refuse_chap(setting);
-  }
-  else if (apn->access_method)
-  {
-    return TRUE;
-  }
-  return chap_enabled;
-}
-
-/**
- * cc_wwan_data_apn_get_initial_eps_refuse_mschap:
- * @apn: A #CcWwanDataApn
- *
- * For LTE modems, get the mschap authentication status of @apn
+ * For LTE modems, get the authentication status for all auth types of @apn
  *
- * Returns: (transfer none): The mschap authentication status of @apn
+ * Returns: (transfer none): The authentication status of @apn
  */
-static gboolean
-cc_wwan_data_apn_get_initial_eps_refuse_mschap (CcWwanDataApn *apn)
+guint
+cc_wwan_data_apn_get_initial_eps_auth_method (CcWwanDataApn *apn)
 {
-  gboolean mschap_enabled;
-
-
-  if (apn->remote_connection)
-  {
-    NMSettingGsm *setting;
-    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
-    mschap_enabled = nm_setting_gsm_get_initial_eps_refuse_mschap(setting);
-  }
-  else if (apn->access_method)
-  {
-    return FALSE;
-  }
-  return mschap_enabled;
-}
+  NMSettingGsm *setting;
 
-/**
- * cc_wwan_data_apn_get_initial_eps_refuse_mschapv2:
- * @apn: A #CcWwanDataApn
- *
- * For LTE modems, get the mschapv2 authentication status of @apn
- *
- * Returns: (transfer none): The mschapv2 authentication status of @apn
- */
-static gboolean
-cc_wwan_data_apn_get_initial_eps_refuse_mschapv2 (CcWwanDataApn *apn)
-{
-  gboolean mschapv2_enabled;
+  g_return_val_if_fail (CC_IS_WWAN_DATA_APN (apn), 0);
 
+  if (!apn->remote_connection)
+    return 0;
 
-  if (apn->remote_connection)
-  {
-    NMSettingGsm *setting;
-    setting = nm_connection_get_setting_gsm(NM_CONNECTION(apn->remote_connection));
-    mschapv2_enabled = nm_setting_gsm_get_initial_eps_refuse_mschapv2(setting);
-  }
-  else if (apn->access_method)
-  {
-    return FALSE;
-  }
-  return mschapv2_enabled;
-}
+  setting = nm_connection_get_setting_gsm (NM_CONNECTION (apn->remote_connection));
+  if (nm_setting_gsm_get_initial_eps_noauth (setting))
+    return 0;
+  if (!nm_setting_gsm_get_initial_eps_refuse_pap (setting))
+    return 1;
+  if (!nm_setting_gsm_get_initial_eps_refuse_chap (setting))
+    return 2;
+  if (!nm_setting_gsm_get_initial_eps_refuse_eap (setting))
+    return 3;
+  if (!nm_setting_gsm_get_initial_eps_refuse_mschap (setting))
+    return 4;
+  if (!nm_setting_gsm_get_initial_eps_refuse_mschapv2 (setting))
+    return 5;
 
-/**
- * cc_wwan_data_apn_get_initial_eps_auth:
- * @apn: A #CcWwanDataApn
- *
- * For LTE modems, get the authentication status for all auth types of @apn
- *
- * Returns: (transfer none): The authentication status of @apn
- */
-guint
-cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn)
-{
-  if(cc_wwan_data_apn_get_initial_eps_noauth(apn) == TRUE) return 0;
-  else if(cc_wwan_data_apn_get_initial_eps_refuse_pap(apn) == FALSE) return 1;
-  else if(cc_wwan_data_apn_get_initial_eps_refuse_chap(apn) == FALSE) return 2;
-  else if(cc_wwan_data_apn_get_initial_eps_refuse_eap(apn) == FALSE) return 3;
-  else if(cc_wwan_data_apn_get_initial_eps_refuse_mschap(apn) == FALSE) return 4;
-  else if(cc_wwan_data_apn_get_initial_eps_refuse_mschapv2(apn) == FALSE) return 5;
-  else return 0;
+  return 0;
 }
 
 /**
- * cc_wwan_data_apn_set_initial_eps_auth:
+ * cc_wwan_data_apn_set_initial_eps_auth_method:
  * @apn: A #CcWwanDataApn
  * @authtype: The initial eps bearer auth type
  *
@@ -1850,8 +1703,8 @@ cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn)
  * to save the changes.
  */
 void
-cc_wwan_data_apn_set_initial_eps_auth ( CcWwanDataApn *apn,
-                            guint authtype)
+cc_wwan_data_apn_set_initial_eps_auth_method (CcWwanDataApn *apn,
+                                              guint authtype)
 {
   NMSettingGsm *setting;
   NMConnection *connection;
diff --git a/panels/wwan/cc-wwan-data.h b/panels/wwan/cc-wwan-data.h
index 1539fc9d6..6644c45b8 100644
--- a/panels/wwan/cc-wwan-data.h
+++ b/panels/wwan/cc-wwan-data.h
@@ -99,9 +99,9 @@ const gchar   *cc_wwan_data_apn_get_initial_eps_password (CcWwanDataApn *apn);
 void           cc_wwan_data_apn_set_initial_eps_password (CcWwanDataApn *apn,
                                                           const gchar   *password);
 
-guint cc_wwan_data_apn_get_initial_eps_auth (CcWwanDataApn *apn);
-void cc_wwan_data_apn_set_initial_eps_auth (CcWwanDataApn *apn,
-                                            guint allowed_auth);
+guint cc_wwan_data_apn_get_initial_eps_auth_method (CcWwanDataApn *apn);
+void cc_wwan_data_apn_set_initial_eps_auth_method  (CcWwanDataApn *apn,
+                                                    guint allowed_auth);
 
 gboolean cc_wwan_data_apn_should_configure_initial_eps_bearer (CcWwanDataApn *apn);
 
-- 
2.49.0


From 786889116e2e3a178f21a9cffb082a8a1e834ffa Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 20 May 2025 11:55:15 +0200
Subject: [PATCH 5/6] wwan, apn-dialog: Only show APN auth-type when APN-type
 is "Attach"

https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/2454#note_2439327
(cherry picked from commit 5d2e971fd1c859529cffd8b2a7b3fcb2fecd9b3a)
---
 panels/wwan/cc-wwan-apn-dialog.ui | 35 ++++++++++++++++---------------
 1 file changed, 18 insertions(+), 17 deletions(-)

diff --git a/panels/wwan/cc-wwan-apn-dialog.ui b/panels/wwan/cc-wwan-apn-dialog.ui
index 558a5ffa3..259605b7a 100644
--- a/panels/wwan/cc-wwan-apn-dialog.ui
+++ b/panels/wwan/cc-wwan-apn-dialog.ui
@@ -193,14 +193,13 @@
                       </object>
                     </child>
 
-
-                    <!--Initial EPS Auth Type -->
+                    <!-- Initial EPS APN type -->
                     <child>
                       <object class="GtkLabel">
                         <property name="halign">end</property>
                         <property name="valign">center</property>
-                        <property name="label" translatable="yes">Auth Type</property>
-                        <property name="mnemonic_widget">lte_auth_combo</property>
+                        <property name="label" translatable="yes">APN Type</property>
+                        <property name="mnemonic_widget">lte_attach_combo</property>
                         <style>
                           <class name="dim-label"/>
                         </style>
@@ -212,16 +211,12 @@
                     </child>
                     
                     <child>
-                      <object class="GtkDropDown" id="lte_auth_combo">
+                      <object class="GtkDropDown" id="lte_attach_combo">
                         <property name="model">
                           <object class="GtkStringList">
                             <items>
-                              <item translatable="yes">None</item>
-                              <item translatable="yes">PAP</item>
-                              <item translatable="yes">CHAP</item>
-                              <item translatable="yes">EAP</item>
-                              <item translatable="yes">MSCHAP</item>
-                              <item translatable="yes">MSCHAPV2</item>
+                              <item translatable="yes">Internet</item>
+                              <item translatable="yes">Attach</item>
                             </items>
                           </object>
                         </property>
@@ -232,13 +227,14 @@
                       </object>
                     </child>
 
-                    <!-- Initial EPS APN type -->
+                    <!--Initial EPS Auth Type -->
                     <child>
                       <object class="GtkLabel">
+                        <property name="visible" bind-source="lte_attach_combo" bind-property="selected" bind-flags="sync-create|bidirectional"/>
                         <property name="halign">end</property>
                         <property name="valign">center</property>
-                        <property name="label" translatable="yes">APN Type</property>
-                        <property name="mnemonic_widget">lte_attach_combo</property>
+                        <property name="label" translatable="yes">Auth Type</property>
+                        <property name="mnemonic_widget">lte_auth_combo</property>
                         <style>
                           <class name="dim-label"/>
                         </style>
@@ -250,12 +246,17 @@
                     </child>
                     
                     <child>
-                      <object class="GtkDropDown" id="lte_attach_combo">
+                      <object class="GtkDropDown" id="lte_auth_combo">
+                        <property name="visible" bind-source="lte_attach_combo" bind-property="selected" bind-flags="sync-create|bidirectional"/>
                         <property name="model">
                           <object class="GtkStringList">
                             <items>
-                              <item translatable="yes">Internet</item>
-                              <item translatable="yes">Attach</item>
+                              <item translatable="yes">None</item>
+                              <item translatable="yes">PAP</item>
+                              <item translatable="yes">CHAP</item>
+                              <item translatable="yes">EAP</item>
+                              <item translatable="yes">MSCHAP</item>
+                              <item translatable="yes">MSCHAPV2</item>
                             </items>
                           </object>
                         </property>
-- 
2.49.0


From 81b50044056bd00f1c4492104cff19fd1a295c48 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 20 May 2025 12:00:01 +0200
Subject: [PATCH 6/6] wwan, apn-dialog: Don't translate authentication protocol
 names

https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/2454#note_2439317
(cherry picked from commit 9921f103a1b28269b102794ac230dc2b10633c80)
---
 panels/wwan/cc-wwan-apn-dialog.ui | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/panels/wwan/cc-wwan-apn-dialog.ui b/panels/wwan/cc-wwan-apn-dialog.ui
index 259605b7a..f86c3ae2c 100644
--- a/panels/wwan/cc-wwan-apn-dialog.ui
+++ b/panels/wwan/cc-wwan-apn-dialog.ui
@@ -252,11 +252,11 @@
                           <object class="GtkStringList">
                             <items>
                               <item translatable="yes">None</item>
-                              <item translatable="yes">PAP</item>
-                              <item translatable="yes">CHAP</item>
-                              <item translatable="yes">EAP</item>
-                              <item translatable="yes">MSCHAP</item>
-                              <item translatable="yes">MSCHAPV2</item>
+                              <item>PAP</item>
+                              <item>CHAP</item>
+                              <item>EAP</item>
+                              <item>MSCHAP</item>
+                              <item>MSCHAPV2</item>
                             </items>
                           </object>
                         </property>
-- 
2.49.0

